### ISSUE TITLE
network: Match.SSID= doesn't work correctly with AP interfaces

### ISSUE THREAD

**Issue by sipraga** [systemd/systemd#25384]
### systemd version the issue has been seen with

develop

### Used distribution

Yocto

### Linux kernel version used

5.18.5

### CPU architectures issue was seen on

aarch64

### Component

~~networkctl~~ networkd

### Expected behaviour you didn't see

networkd correctly applies the Match directive SSID=foo when a wireless interface is configured via hostapd with SSID "foo".

### Unexpected behaviour you saw

networkd considers the SSID unset, and hence a .network configuration with `Match.SSID=foo` will never match on the wireless interface.

### Steps to reproduce the problem

Consider the following trivial networkd config:
```
[Match]
SSID=foo

[Network]
Address=172.18.128.1
```

Start networkd with such a configuration available. Then start hostapd on a network interface `uap0` with the following configuration file (`hostapd ./hostapd.conf`):
```
interface=uap0
ssid=foo
hw_mode=g
channel=1
```

Observe that networkd does not detect any change in `uap0`'s SSID and hence the above network configuration is never applied to `uap0`.

**Comment by yuwata** [comment]
Duplicate of #24585, and this is a kernel bug.

**Comment by sipraga** [comment]
Hi @yuwata, I have read #24585 but this is a different bug. Note that I am using 5.18 but the supposed kernel regression appears in 5.19 according to the other bug report. And anyway, I don't have any issue with SSID on station interfaces.

Please see a more detailed description below and consider re-opening this issue:

The problem here is that networkd only has infrastructure to detect SSID in the following scenarios:
1. when networkd first starts and enumerates available network interfaces (but then the AP might not be started yet, so this is insufficient)
2. in response to a netlink multicast message `NL80211_CMD_CONNECT` or `NL80211_CMD_DISCONNECT` (but these messages are only broadcast by the kernel for station iftypes, not APs, so this is also insufficient)

Regarding (2), regrettably the kernel doesn't broadcast any nl80211 multicast group message when a BSS is started or stopped on an AP interface. AFAIK the only thing that networkd has to react to is the carrier state (and networkd actually used to do this some versions ago before refactoring was done to handle wireless station roaming better). When gaining carrier, networkd can issue a CMD_GET_INTERFACE to retrieve the SSID.

**Comment by sipraga** [comment]
I think I made a mistake with selecting "Component: networkctl" in the original report above. This is a networkd issue. networkctl has no bug.

**Comment by yuwata** [comment]
@sipraga Fix is waiting in #25477. If possible, could you test the PR?

**Comment by sipraga** [comment]
@yuwata thanks for following up. I tested the PR but it does not solve my issue. The problem is that there aren't any nl80211 events when an AP goes up and down - you only see these events for stations. I think that the SSID needs to be re-parsed in `link_carrier_gained()`, as this is the only event we get from the kernel signalling that an AP has been enabled on an interface.

This is my crude fix to the problem: https://github.com/systemd/systemd/pull/25488
I hope you get the idea.

Not sure if it's the correct way, but it seems to solve the problem on my side. We can discuss further on that PR if you like.

**Comment by yuwata** [comment]
Fixed by #24588.
